### PR TITLE
Add Chass Import Button

### PR DESCRIPTION
**Pull request by michellemtchai**
- implemented `/import/chass` route
- modified `ChassImporter` so that no file is saved onto the server for import. The upload file in passed as a json to the server not as a file
- updated rspec for `ChassImporter` accordingly
- added info button next to upload text
  - should show a popup with the CHASS JSON format
- implemented `upload` button on Summary view
- The following are the alerts you may see when trying to press the upload button:
  - `Error: No file chosen.`
  - `Are you sure you want to import "<filename>" into the database?`
  - `Error: The file you uploaded is not a JSON.`
  - `Import Failure: no round_id found in the file.`
  - `Import Failure: too many round_id found in the file.`

### CODE REVIEW

File: app/assets/stylesheets/app.css [hunk 1]
```diff
@@ -425,4 +449,34 @@ footer div p {
...
+    padding: 0;
+    border: none;
+    background: none;
+    outline: 0;
+}
+#help {
+    top: 76px !important;
+}
+#help .popover-content {
+    padding: 0 !important;
```
Comment (gabriellesc) [inline app/assets/stylesheets/app.css]: I think it looks a little better with some padding - maybe 5px?

File: app/assets/stylesheets/app.css [hunk 2]
```diff
@@ -425,4 +449,34 @@ footer div p {
...
+    width: 300px;
+    height: 200px;
```
Comment (gabriellesc) [inline app/assets/stylesheets/app.css]: I think it's a bit easier to read if it's taller - maybe 400px?

File: app/assets/stylesheets/app.css [hunk 3]
```diff
@@ -425,4 +449,34 @@ footer div p {
...
+}
+#help .popover-content textarea {
+    width: calc(100% - 27px);
```
Comment (gabriellesc) [inline app/assets/stylesheets/app.css]: Just wanted to make a note that, if padding is added, this width calculation will change...

File: app/javascript/app/components/summary.js [hunk 4]
```diff
@@ -10,7 +10,11 @@ import {
     Button,
     Well,
     Table,
+    OverlayTrigger,
+    Overlay,
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: `Overlay` doesn't appear to be used anywhere?

File: app/controllers/import_controller.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+require 'future'
+class ImportController < ApplicationController
+  protect_from_forgery with: :null_session
+
+  def chass
+    import = promise{ChassImporter.new(params[:chass_json])}
```
Comment (freeatnet) [inline app/controllers/import_controller.rb]: Hmm. Could you explain what happens here?

File: app/javascript/app/components/summary.js [hunk 6]
```diff
@@ -10,7 +10,11 @@ import {
...
+    Popover,
 } from 'react-bootstrap';
+import * as fetch from '../fetch.js';
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: fetch should *not* be imported here - any fetch functions should be wrapped in appState, and the appState wrapper function should be called from here

File: package.json [hunk 7]
```diff
@@ -19,6 +19,7 @@
     "css-loader": "^0.28.2",
     "extract-text-webpack-plugin": "^2.1.0",
     "file-loader": "^0.11.1",
+    "gem": "^2.4.2",
```
Comment (freeatnet) [inline package.json]: What's this for?

File: app/javascript/app/components/summary.js [hunk 8]
```diff
@@ -45,12 +49,41 @@ class Summary extends React.Component {
...
         this.selectThisTab();
     }
+
+    loadFile() {
+        let files = document.getElementById('import').files;
+        if (files.length > 0) {
+            let message =
+                'Are you sure you want to import "' + files[0].name + '" into the database?';
+            if (files[0].type == 'application/json') {
+                if (confirm(message)) this.uploadFile(files[0]);
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: Missing curly braces, so `prettier` put this on a single line...

File: app/javascript/app/components/summary.js [hunk 9]
```diff
@@ -45,12 +49,41 @@ class Summary extends React.Component {
     componentWillUpdate() {
         this.selectThisTab();
     }
+
+    loadFile() {
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: There's no reason to put `loadFile` and `uploadFile` in `Summary` - I think it makes more sense to put them directly into `ImportForm`. I've turned `ImportForm` and `ExportForm` into more than functional classes in #126, where member functions can be attached.

File: app/javascript/app/fetch.js [hunk 10]
```diff
@@ -411,11 +411,21 @@ function updateCourse(courseId, data, val, attr) {
         .catch(() => appState.setFetchingCoursesList(false));
 }
 
+function importChass(data, success, failure) {
+    return postHelper('/import/chass', data, success, failure);
+}
+
+function showMessage(resp) {
```
Comment (gabriellesc) [inline app/javascript/app/fetch.js]: I think we should stick to the alert and notification functionality that we've defined; in this case, I think it makes sense to make this a notification.

File: app/javascript/app/fetch.js [hunk 11]
```diff
@@ -411,11 +411,21 @@ function updateCourse(courseId, data, val, attr) {
         .catch(() => appState.setFetchingCoursesList(false));
 }
 
+function importChass(data, success, failure) {
+    return postHelper('/import/chass', data, success, failure);
```
Comment (gabriellesc) [inline app/javascript/app/fetch.js]: We need to call fetchAll after this completes successfully!

File: app/javascript/app/components/summary.js [hunk 12]
```diff
@@ -59,16 +92,37 @@ const Utilities = props => {
...
+        <FormGroup>
+            <ControlLabel>
+                Import from file:&nbsp;
+                <OverlayTrigger
+                    trigger="click"
+                    rootClose
+                    placement="right"
+                    overlay={InfoDialog(chassFormat)}
+                >
+                    <button className="info-button" type="button">
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: I don't think that this icon needs to be wrapped in a button?

File: app/javascript/app/components/summary.js [hunk 13]
```diff
@@ -95,7 +150,8 @@ const ExportForm = props =>
                 componentClass="select"
                 inputRef={ref => {
                     this.format = ref;
-                }}>
+                }}
+            >
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: This is a minor thing but it seems really strange to me - when I run `prettier` on this, it actually puts the closing `>` on the same line as the `}}`. I'm wondering if your `prettier` is configured differently from https://github.com/uoft-tapp/tapp/wiki/JavaScript-JSX-Style ?

**Review by gabriellesc — CHANGES_REQUESTED** [review]

File: app/controllers/import_controller.rb (hunk 5 above)
Comment (michellemtchai) [reply app/controllers/import_controller.rb]: This waits for the ChassImporter to finish inserting the CHASS JSON to the DB before `importer.done` is called for the Import complete message. Before that no render is done by the controller.

File: app/javascript/app/components/summary.js (hunk 13 above)
Comment (michellemtchai) [reply app/javascript/app/components/summary.js]: It is actually the same setting as what's on the link. I don't know why it's causing this issue.

File: app/controllers/import_controller.rb (hunk 5 above)
Comment (freeatnet) [reply app/controllers/import_controller.rb]: Are there any asynchronous operations in ChassImporter? If there is not, maybe we don't need an asynchronous concept of promises here.

Comment (michellemtchai) [reply app/controllers/import_controller.rb]: When I try to do without the promise, the render always finishes before the import is done.

**Review by jmzaleski — COMMENTED** [review]
worked on my machine!

File: app/javascript/app/components/summary.js [hunk 14]
```diff
@@ -25,7 +27,7 @@ class Summary extends React.Component {
         return (
             <Grid fluid id="summary-grid" style={cursorStyle}>
                 <PanelGroup>
-                    <Utilities {...this.props} />
+                    <Utilities loadFile={() => this.loadFile()} {...this.props} />
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: I think this should be removed?

File: app/javascript/app/components/summary.js [hunk 15]
```diff
@@ -50,30 +52,80 @@ class Summary extends React.Component {
 const Utilities = props => {
     return (
         <Panel header="Utilities" id="utils">
-            <ImportForm {...props} />
+            <ImportForm loadFile={props.loadFile} {...props} />
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: I think this should be removed?

**Review by jmzaleski — APPROVED** [review]

File: app/javascript/app/components/summary.js [hunk 16]
```diff
@@ -50,30 +52,80 @@ class Summary extends React.Component {
...
+                }
+            } else {
+                alert('Error: The file you uploaded is not a JSON.');
+            }
+        } else {
+            alert('Error: No file chosen.');
+        }
+    }
+
+    uploadFile(file, func) {
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: Would it be possible to give `func` a more descriptive name and/or put in a comment describing its purpose?

File: app/javascript/app/components/summary.js [hunk 17]
```diff
@@ -50,30 +52,80 @@ class Summary extends React.Component {
...
+        let reader = new FileReader();
+        reader.onload = function(event) {
+            let data = JSON.parse(event.target.result);
+            if (data['courses'] !== undefined && data['applicants'] !== undefined) {
+                data = { chass_json: data };
+                func(data);
+            } else {
+                alert('Error: This is not a CHASS JSON.');
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: I think we should use the app alert function here?

File: app/javascript/app/components/summary.js [hunk 18]
```diff
@@ -50,30 +52,80 @@ class Summary extends React.Component {
...
+                'Are you sure you want to import "' + files[0].name + '" into the database?';
+            if (files[0].type == 'application/json') {
+                if (confirm(message)) {
+                    this.uploadFile(files[0], this.props.func.importChass);
+                }
+            } else {
+                alert('Error: The file you uploaded is not a JSON.');
+            }
+        } else {
+            alert('Error: No file chosen.');
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: I think we should use the app alert function here?

File: app/javascript/app/components/summary.js [hunk 19]
```diff
@@ -50,30 +52,80 @@ class Summary extends React.Component {
...
+        let files = document.getElementById('import').files;
+        if (files.length > 0) {
+            let message =
+                'Are you sure you want to import "' + files[0].name + '" into the database?';
+            if (files[0].type == 'application/json') {
+                if (confirm(message)) {
+                    this.uploadFile(files[0], this.props.func.importChass);
+                }
+            } else {
+                alert('Error: The file you uploaded is not a JSON.');
```
Comment (gabriellesc) [inline app/javascript/app/components/summary.js]: I think we should use the app alert function here?

**Review by gabriellesc — APPROVED** [review]

**Pull request merged**